README.md
# Cloudability - Money

## Contributing

See the [Contribution Guidelines](https://github.com/RubyMoney/money/blob/master/CONTRIBUTING.md)

## Introduction

This library aids one in handling money and different currencies.

### Features

- Provides a `Money` class which encapsulates all information about an certain
  amount of money, such as its value and its currency.
- Provides a `Money::Currency` class which encapsulates all information about
  a monetary unit.
- Represents monetary values as integers, in cents. This avoids floating point
  rounding errors.
- Represents currency as `Money::Currency` instances providing an high level of
  flexibility.
- Provides APIs for exchanging money from one currency to another.
- Has the ability to parse a money and currency strings
  into the corresponding Money/Currency object.

### Resources

- [Website](http://rubymoney.github.com/money)
- [API Documentation](http://rubydoc.info/gems/money/frames)
- [Git Repository](http://github.com/RubyMoney/money)

### Notes

- Your app must use UTF-8 to function with this library. There are a
  number of non-ASCII currency attributes.
- This app requires Ruby 1.9 and JSON. If you're using JRuby < 1.7.0
  you'll need to add `gem "json"` to your Gemfile or similar.

## Downloading

Install stable releases with the following command:

    gem install money

The development version (hosted on Github) can be installed with:

    git clone git://github.com/RubyMoney/money.git
    cd money
    rake install

## Usage

``` ruby
require 'money'

# 10.00 USD
money = Money.new(1000, "USD")
money.cents     #=> 1000
money.currency  #=> Currency.new("USD")

# Comparisons
Money.new(1000, "USD") == Money.new(1000, "USD")   #=> true
Money.new(1000, "USD") == Money.new(100, "USD")    #=> false
Money.new(1000, "USD") == Money.new(1000, "EUR")   #=> false
Money.new(1000, "USD") != Money.new(1000, "EUR")   #=> true

# Arithmetic
Money.new(1000, "USD") + Money.new(500, "USD") == Money.new(1500, "USD")
Money.new(1000, "USD") - Money.new(200, "USD") == Money.new(800, "USD")
Money.new(1000, "USD") / 5                     == Money.new(200, "USD")
Money.new(1000, "USD") * 5                     == Money.new(5000, "USD")

# Assumptive Currencies
Money.assume_from_symbol = true
Money.parse("$100") == Money.new(10000, "USD")
Money.parse("€100") == Money.new(10000, "EUR")
Money.parse("£100") == Money.new(10000, "GBP")

# Currency conversions
some_code_to_setup_exchange_rates
Money.new(1000, "USD").exchange_to("EUR") == Money.new(some_value, "EUR")
```

## Currency

Currencies are consistently represented as instances of `Money::Currency`.
The most part of `Money` APIs allows you to supply either a `String` or a
`Money::Currency`.

``` ruby
Money.new(1000, "USD") == Money.new(1000, Currency.new("USD"))
Money.new(1000, "EUR").currency == Currency.new("EUR")
```

A `Money::Currency` instance holds all the information about the currency,
including the currency symbol, name and much more.

``` ruby
currency = Money.new(1000, "USD").currency
currency.iso_code #=> "USD"
currency.name     #=> "United States Dollar"
```

To define a new `Money::Currency` use `Money::Currency.register` as shown
below.

``` ruby
curr = {
  :priority        => 1,
  :iso_code        => "USD",
  :iso_numeric     => "840",
  :name            => "United States Dollar",
  :symbol          => "$",
  :subunit         => "Cent"
  :subunit_to_unit => 100,
  :separator       => ".",
  :delimiter       => ","
}

Money::Currency.register(curr)
```

The pre-defined set of attributes includes:

- `:priority` a numerical value you can use to sort/group the currency list
- `:iso_code` the international 3-letter code as defined by the ISO 4217 standard
- `:iso_numeric` the international 3-digit code as defined by the ISO 4217 standard
- `:name` the currency name
- `:symbol` the currency symbol (UTF-8 encoded)
- `:subunit` the name of the fractional monetary unit
- `:subunit_to_unit` the proportion between the unit and the subunit
- `:separator` character between the whole and fraction amounts
- `:delimiter` character between each thousands place

All attributes are optional. Some attributes, such as `:symbol`, are used by
the Money class to print out a representation of the object. Other attributes,
such as `:name` or `:priority`, exist to provide a basic API you can take
advantage of to build your application.

### :priority

The priority attribute is an arbitrary numerical value you can assign to the
`Money::Currency` and use in sorting/grouping operation.

For instance, let's assume your Rails application needs to render a currency
selector like the one available
[here](http://finance.yahoo.com/currency-converter/). You can create a couple of
custom methods to return the list of major currencies and all currencies as
follows:

``` ruby
# Returns an array of currency id where
# priority < 10
def major_currencies(hash)
  hash.inject([]) do |array, (id, attributes)|
    priority = attributes[:priority]
    if priority && priority < 10
      array[priority] ||= []
      array[priority] << id
    end
    array
  end.compact.flatten
end

# Returns an array of all currency id
def all_currencies(hash)
  hash.keys
end

major_currencies(Money::Currency.table)
# => [ :usd, :eur, :bgp, :cad ]

all_currencies(Money::Currency.table)
# => [ :aed, :afn, all, ... ]
```

### Default Currency

By default `Money` defaults to USD as its currency. This can be overwritten
using:

``` ruby
Money.default_currency = Money::Currency.new("CAD")
```

If you use Rails, then `environment.rb` is a very good place to put this.

## Currency Exchange

Exchanging money is performed through an exchange bank object. The default
exchange bank object requires one to manually specify the exchange rate. Here's
an example of how it works:

``` ruby
Money.add_rate("USD", "CAD", 1.24515)
Money.add_rate("CAD", "USD", 0.803115)

Money.us_dollar(100).exchange_to("CAD")  # => Money.new(124, "CAD")
Money.ca_dollar(100).exchange_to("USD")  # => Money.new(80, "USD")
```

Comparison and arithmetic operations work as expected:

``` ruby
Money.new(1000, "USD") <=> Money.new(900, "USD")   # => 1; 9.00 USD is smaller
Money.new(1000, "EUR") + Money.new(10, "EUR") == Money.new(1010, "EUR")

Money.add_rate("USD", "EUR", 0.5)
Money.new(1000, "EUR") + Money.new(1000, "USD") == Money.new(1500, "EUR")
```

There is nothing stopping you from creating bank objects which scrapes
[XE](http://www.xe.com) for the current rates or just returns `rand(2)`:

``` ruby
Money.default_bank = ExchangeBankWhichScrapesXeDotCom.new
```

### Implementations

The following is a list of Money.gem compatible currency exchange rate
implementations.

- [eu_central_bank](http://github.com/RubyMoney/eu_central_bank)
- [google_currency](http://github.com/RubyMoney/google_currency)
- [nordea](https://github.com/k33l0r/nordea)
- [nbrb_currency](https://github.com/slbug/nbrb_currency)
- [money-open-exchange-rates](https://github.com/spk/money-open-exchange-rates)
- [money-historical-bank](https://github.com/atwam/money-historical-bank)

## Ruby on Rails

To integrate money in a Rails application use [money-rails](http://github.com/RubyMoney/money-rails).

For depreceated methods of integrating with Rails, check [the wiki](https://github.com/RubyMoney/money/wiki).

CHANGELOG.md
# Changelog

## master

NOTICE: This change breaks the magic hand-waving about commas-vs-periods as
thousands separators vs. decimal separators.  Now, the currency's configuration
will always be respected in this regard.  This is necessary to properly handle
high-precision values.

- Add support for retaining high-precision versions of money values.
- Reformat changelog to actually be Markdown.
- Add a numeric ID for currencies that can fit in an unsigned 1-byte int, handy
  for dimension columns in ROLAP tables.  Addded the `assign_ids` rake task to
  help keep things up to date.

## upstream

- Ensure BigDecimal.new always receives a string - compatibility fix for ruby-1.9.2-p320
- Update Maldivian Currency to MVR and fix ރ. to be ރ

## 5.1.0

- Fix currency assumption when parsing $ with a non-USD default currency.
- Changed the Bulgarian lev symbol position from before the amount to after the amount.
- Changed the symbol and html entity for INR. It is now "₹" instead of "₨".
- Added Money::Currency.analyze for determining potential currencies for a given string using powereful algorithms - will detect symbols, iso codes and names even if mixed with text.
- Changed UGX symbol from 'Sh' to 'USh'.
- Changed SYP symbol from "£ or ل.س" to "£S". The previous symbols remain as alternates.
- Changed RWF symbol from 'FR' to 'FRw'.
- Changed RSD symbol from "din. or дин." to 'РСД'. The previous symbols remain as alternates.
- Added MGA symbol 'Ar'
- Added KGS symbol 'som'
- Changed KES symbol from 'Sh' to 'KSh'
- Added ETB symbol 'Br'
- Changed EGP symbol from "£ or ج.م" to "ج.م"
- Changed DJF symbol from 'Fr' to 'Fdj'
- Changed CVE symbol from '$ or Esc' to '$'. 'Esc' remains as an alternate symbol.
- Added BTN symbol 'Nu.'
- Changed BAM symbol from 'KM or КМ' to 'КМ', the alternate (cyrillic script) remains as an alternate symbol.
- Added alternate symbols for all currencies. For example, USD can be written as both '$' and 'US$', SEK can be 'Kr' or ':-', etc.
- Renamed Money#cents to Money#fractional. Money#cents can still be used as a synonym and equivalent of Money#fractional.
- Added Money.new_with_amount and Money#amount. Money.new_with_dollars and Money#dollars remain as synonyms.
- Calling Bank::Base.instance doesn't make Bank::VariableExchange.instance
  return Bank::Base.instance anymore (semaperepelitsa)
- Update Turkmenistan manat from TMM to TMT currency (GH-181). [Thanks @Exoth]
- Moved ZWD Zimbabwean dollars to currency_bc.json, also added there ZWL, ZWN, and ZWR Zimbabwean dollars (GH-184).
- Switch to multi_json gem (GH-170)
- Fix "warning: ambiguous first argument..." (GH-166)
- Update dependencies to latest and greatest (GH-172)
- TravisBot is now watching Pull Request!!! (GH-171)
- Minor code cleaning
- Remove subunit from South Korean won (KRW)

## 5.0.0

- Minor bugfix - incorrect use of character range resulted in
  botched results for Money::Parsing#extract_cents (GH-162)
- Money::Currency::TABLE removed. Use Money::Currency.register to add
  additional currencies (GH-143)
- Fix rounding error in Numeric.to_money (GH-145)
- Allow on-the-fly calculation of decimal places if not known already
  (GH-146,GH-147,GH-148)
- Move Euro symbol ahead of amount (GH-151)
- Changed settings for Polish currency (GH-152)
- Fall back to symbol if there is no html_entity present (GH-153)
- Optionally Allow parsing of money values prefixed by symbols in key
  currencies (GH-155)
- Fix bug where rates exported to a file from VariableExchange leave the File
  object open (GH-154)
- Added Money#positive? and Money#negative? methods (GH-157)
- Fix format function output for custom currencies (GH-156)
- Fix parsing of strings with 3 decimal digits (GH-158)
- Updated development dependencies
- Said goodbye to RubyForge


## 4.0.2

- Money.to_money now understands a currency option (GH-121)
- Added Money#-@ method to change object polarity (GH-122)
- Added Money#symbol_or_iso_code utility method (GH-128)
- Money.parse now understands trailing - as negative inputs (GH-133)
- Money::Currency.new now validates input to avoid memory leaks (GH-137)

- Forced UTF-8 encoding on currency JSON (GH-117)
- Fixed position of Philippine peso sign (GH-124)
- Fixed position of Danish currency sign (GH-127)


## 4.0.1

- Add missing config dir.


## 4.0.0

The big change this release is moving the currency information into a JSON
file. This should make it easier for users to add and test things. Thanks to
Steve Morris for working on this during his Mendicant University course.

In addition to this big change there were well over a dozen other minor
changes.

### Features

- a new exchange bank nordea has been added to the README. (k33l0r)
- a new exchange bank nbrb_currency has been added to the README. (slbug)
- update Rake tasks
- See our CI status in the README
- Add syntax highlighting to the README (phlipper)
- Remove minor unit from YEN (pwim)
- Format YEN (pwim)
- Update README for `_as_string` (mjankowski)
- Update Lebanon currency (kaleemullah)
- Update Polish złoty (holek)
- Move currency information into JSON storage! (stevemorris)
- Add ISO4217 Numeric codes (alovak)

### Bugfixes

- EEK currency is no longer used, kept for BC ([#issue/110](http://github.com/RubyMoney/money/issues/110))
- Lithuanian Litas symbol position fixed (laurynas)
- Fixed README typos (phlipper)
- Fixed README typos (pwim)
- Fix specs (alovak)


## 3.7.1

### Bugfixes

- Add encoding indicator to top of Gemspec


## 3.7.0

### Features

- add Money#to_d (thanks Andrew White)
- Add Money.use_i18n, this allows you to enable/disable i18n from being used,
  even if it's required in your app.


## 3.6.2

### Features

- i18n enhancements (thanks eloyesp [link](https://github.com/RubyMoney/money/commit/b2cab76c78ae04f40251fa20c4ab18faa968dc53))
- README updates (thanks pconnor)
- Break into modules
- Added `:no_cents_if_whole` format option
- Update HKD from Ho to Cent
- Performance improvements (thanks weppos)

spec/currency_spec.rb
# encoding: utf-8

require "spec_helper"

describe Money::Currency do

  FOO = '{ "priority": 1, "iso_code": "FOO", "iso_numeric": "840", "name": "United States Dollar", "symbol": "$", "subunit": "Cent", "subunit_to_unit": 450, "symbol_first": true, "html_entity": "$", "decimal_mark": ".", "thousands_separator": "," }'

  describe ".find" do
    it "returns currency matching given id" do
      Money::Currency.register(JSON.parse(FOO, :symbolize_names => true))

      expected = Money::Currency.new(:foo)
      Money::Currency.find(:foo).should  == expected
      Money::Currency.find(:FOO).should  == expected
      Money::Currency.find("foo").should == expected
      Money::Currency.find("FOO").should == expected
    end

    it "returns nil unless currency matching given id" do
      Money::Currency.find("ZZZ").should be_nil
    end
  end

  describe ".wrap" do
    it "returns nil if object is nil" do
      Money::Currency.wrap(nil).should == nil
      Money::Currency.wrap(Money::Currency.new(:usd)).should == Money::Currency.new(:usd)
      Money::Currency.wrap(:usd).should == Money::Currency.new(:usd)
    end
  end


  describe "#initialize" do
    it "lookups data from loaded config" do
      currency = Money::Currency.new("USD")
      currency.key.should                 == :usd
      currency.priority.should            == 1
      currency.iso_code.should            == "USD"
      currency.iso_numeric.should         == "840"
      currency.name.should                == "United States Dollar"
      currency.decimal_mark.should        == "."
      currency.separator.should           == "."
      currency.thousands_separator.should == ","
      currency.delimiter.should           == ","
    end

    it "raises UnknownMoney::Currency with unknown currency" do
      expect { Money::Currency.new("xxx") }.to raise_error(Money::Currency::UnknownCurrency, /xxx/)
    end
  end

  describe "#<=>" do
    it "compares objects by priority" do
      Money::Currency.new(:cad).should > Money::Currency.new(:usd)
      Money::Currency.new(:usd).should < Money::Currency.new(:eur)
    end
  end

  describe "#==" do
    it "returns true if self === other" do
      currency = Money::Currency.new(:eur)
      currency.should == currency
    end

    it "returns true if the id is equal" do
      Money::Currency.new(:eur).should     == Money::Currency.new(:eur)
      Money::Currency.new(:eur).should_not == Money::Currency.new(:usd)
    end
  end

  describe "#eql?" do
    it "returns true if #== returns true" do
      Money::Currency.new(:eur).eql?(Money::Currency.new(:eur)).should be true
      Money::Currency.new(:eur).eql?(Money::Currency.new(:usd)).should be false
    end
  end

  describe "#hash" do
    it "returns the same value for equal objects" do
      Money::Currency.new(:eur).hash.should == Money::Currency.new(:eur).hash
      Money::Currency.new(:eur).hash.should_not == Money::Currency.new(:usd).hash
    end

    it "can be used to return the intersection of Currency object arrays" do
      intersection = [Money::Currency.new(:eur), Money::Currency.new(:usd)] & [Money::Currency.new(:eur)]
      intersection.should == [Money::Currency.new(:eur)]
    end
  end

  describe "#inspect" do
    it "works as documented" do
      Money::Currency.new(:usd).inspect.should ==
          %Q{#<Money::Currency id: 146, key: usd, priority: 1, symbol_first: true, thousands_separator: ,, html_entity: $, decimal_mark: ., name: United States Dollar, symbol: $, subunit_to_unit: 100, iso_code: USD, iso_numeric: 840, subunit: Cent>}
    end
  end

  describe "#to_s" do
    it "works as documented" do
      Money::Currency.new(:usd).to_s.should == "USD"
      Money::Currency.new(:eur).to_s.should == "EUR"
    end
  end

  describe "#to_currency" do
    it "works as documented" do
      usd = Money::Currency.new(:usd)
      usd.to_currency.should == usd
    end

    it "doesn't create new symbols indefinitely" do
      expect { Money::Currency.new("bogus") }.to raise_exception(Money::Currency::UnknownCurrency)
      Symbol.all_symbols.map{|s| s.to_s}.should_not include("bogus")
    end
  end

  describe "#code" do
    it "works as documented" do
      Money::Currency.new(:usd).code.should == "$"
      Money::Currency.new(:azn).code.should == "AZN"
    end
  end

  describe "#decimal_places" do
    it "proper places for known currency" do
      Money::Currency.new(:mro).decimal_places == 1
      Money::Currency.new(:usd).decimal_places == 2
    end

    it "proper places for custom currency" do
      Money::Currency.register(JSON.parse(FOO, :symbolize_names => true))
      Money::Currency.new(:foo).decimal_places == 3
    end
  end
end


lib/money/currency/loader.rb
module Money::Currency::Loader
  extend self

  DATA_PATH = File.expand_path("../../../../config", __FILE__)

  # Loads and returns the currencies stored in JSON files in the config directory.
  #
  # @return [Hash]
  def load_currencies(silence_missing_ids_error = false)
    currencies = parse_currency_file("currency.json")
    currencies.merge! parse_currency_file("currency_bc.json")
    currencies.each do |key, currency|
      if(!currency.has_key?(:key))
        currency[:key] = key
      else
        raise Exception.new("Currency already has key (#{currency[:key]}) but we want to give it a new one (#{key})!")
      end
    end

    begin
      parse_currency_file("currency_ids.json").
        each_pair { |key, id| currencies[key][:id] = id }
      missing_ids = currencies.values.select { |currency| !currency.has_key?(:id) }

      if(missing_ids.count > 0)
        missing_keys = missing_ids.map { |currency| currency[:iso_code] }.join(', ')
        raise Exception.new("The following currencies are missing short IDs.  Go run 'rake assign_ids' to fix it: #{missing_keys}")
      end
    rescue Exception => e
      unless(silence_missing_ids_error)
        raise e
      end
    end

    return currencies
  end

  private

  def parse_currency_file(filename)
    json = File.read("#{DATA_PATH}/#{filename}")
    json.force_encoding(::Encoding::UTF_8) if defined?(::Encoding)
    JSON.parse(json, :symbolize_names => true)
  end
end


lib/money/currency.rb
# encoding: utf-8

require 'json'

class Money

  # Represents a specific currency unit.
  class Currency
    include Comparable

    require "money/currency/loader"
    extend  Loader

    require "money/currency/heuristics"
    extend  Heuristics

    # Thrown when an unknown currency is requested.
    class UnknownCurrency < StandardError; end

    class << self

      # Lookup a currency with given +id+/+key+ an returns a +Currency+
      # instance on success, +nil+ otherwise.
      #
      # @param [String, Symbol, Fixnum, #to_s] id/key Used to look into
      # +table+ or +table_by_id+ and retrieve the applicable attributes.
      #
      # @return [Money::Currency]
      #
      # @example
      #   Money::Currency.find(:eur) #=> #<Money::Currency id: 45, key: eur ...>
      #   Money::Currency.find(:foo) #=> nil
      def find(id)
        if(id.is_a?(Fixnum))
          new(id) if self.table_by_id[id]
        else
          id = id.to_s.downcase.to_sym
          new(id) if self.table[id]
        end
      end

      # Wraps the object in a +Currency+ unless it's already a +Currency+
      # object.
      #
      # @param [Object] object The object to attempt and wrap as a +Currency+
      # object.
      #
      # @return [Money::Currency]
      #
      # @example
      #   c1 = Money::Currency.new(:usd)
      #   Money::Currency.wrap(nil)   #=> nil
      #   Money::Currency.wrap(c1)    #=> #<Money::Currency id: 146, key: usd ...>
      #   Money::Currency.wrap("usd") #=> #<Money::Currency id: 146, key: usd ...>
      def wrap(object)
        if object.nil?
          nil
        elsif object.is_a?(Currency)
          object
        else
          Currency.new(object)
        end
      end

      # List of known currencies.
      #
      # == monetary unit
      # The standard unit of value of a currency, as the dollar in the United States or the peso in Mexico.
      # http://www.answers.com/topic/monetary-unit
      # == fractional monetary unit, subunit
      # A monetary unit that is valued at a fraction (usually one hundredth) of the basic monetary unit
      # http://www.answers.com/topic/fractional-monetary-unit-subunit
      #
      # See http://en.wikipedia.org/wiki/List_of_circulating_currencies and
      # http://search.cpan.org/~tnguyen/Locale-Currency-Format-1.28/Format.pm
      def table
        @table ||= load_currencies
      end

      def table_by_id
        @table_by_id ||= Hash[self.table.values.map { |currency| [currency[:id], currency]}]
      end

      # We need a string-based validator before creating an unbounded number of symbols.
      # http://www.randomhacks.net/articles/2007/01/20/13-ways-of-looking-at-a-ruby-symbol#11
      # https://github.com/RubyMoney/money/issues/132
      def stringified_keys
        @stringified_keys ||= stringify_keys
      end

      def register(curr)
        # TODO: What do we do about this case, and numeric IDs?  Hrm...
        key = curr[:iso_code].downcase.to_sym
        @table[key] = curr
        @stringified_keys = stringify_keys
      end

      private

      def stringify_keys
        table.keys.map{|k| k.to_s.downcase}
      end
    end

    # A short, numeric value we assign to currencies so they can be referenced
    # very tersely in places where that matters.  It will not exceed 2^8-1.
    #
    # The +key+ below is not always unique -- several values can correspond to
    # the same +iso_code+, some may not have an +iso_numeric+, etc.  This
    # value will be identical for all Currency objects that have the same
    # iso_code, regardless of the uniqueness (or lack thereof) of the +key+.
    #
    # You should consider very carefully what it means for currencies to be
    # 'equivalent' when deciding which field to use!
    #
    # For example, compare 'JPY' and 'YEN'.
    #
    # @return [Fixnum]
    attr_reader :id

    # The symbol used to identify the currency, usually the lowercase
    # +iso_code+ attribute.
    #
    # @return [Symbol]
    attr_reader :key

    # A numerical value you can use to sort/group the currency list.
    #
    # @return [Integer]
    attr_reader :priority

    # The international 3-letter code as defined by the ISO 4217 standard.
    #
    # @return [String]
    attr_reader :iso_code
    #
    # The international 3-numeric code as defined by the ISO 4217 standard.
    #
    # @return [String]
    attr_reader :iso_numeric

    # The currency name.
    #
    # @return [String]
    attr_reader :name

    # The currency symbol (UTF-8 encoded).
    #
    # @return [String]
    attr_reader :symbol

    # The html entity for the currency symbol
    #
    # @return [String]
    attr_reader :html_entity

    # The name of the fractional monetary unit.
    #
    # @return [String]
    attr_reader :subunit

    # The proportion between the unit and the subunit
    #
    # @return [Integer]
    attr_reader :subunit_to_unit

    # The decimal mark, or character used to separate the whole unit from the subunit.
    #
    # @return [String]
    attr_reader :decimal_mark
    alias :separator :decimal_mark

    # The character used to separate thousands grouping of the whole unit.
    #
    # @return [String]
    attr_reader :thousands_separator
    alias :delimiter :thousands_separator

    # Should the currency symbol precede the amount, or should it come after?
    #
    # @return [boolean]
    attr_reader :symbol_first

    # Create a new +Currency+ object.
    #
    # @param [String, Symbol, Fixnum, #to_s] id Used to look into +table+ or
    #  +table_by_id+ and retrieve the applicable attributes.
    #
    # @return [Money::Currency]
    #
    # @example
    #   Money::Currency.new(:usd) #=> #<Money::Currency id: 146, key: usd ...>
    #   Money::Currency.new(146)  #=> #<Money::Currency id: 146, key: usd ...>
    def initialize(id)
      if(id.is_a?(Fixnum))
        data = self.class.table_by_id[id]
        raise(UnknownCurrency, "Unknown currency ##{id}") unless(data)
      else
        id = id.to_s.downcase
        raise(UnknownCurrency, "Unknown currency `#{id}'") unless self.class.stringified_keys.include?(id)
        data = self.class.table[id.to_sym]
      end
      @id = data[:id]
      @key = data[:key]

      data.each_pair do |key, value|
        instance_variable_set(:"@#{key}", value)
      end
    end

    # Compares +self+ with +other_currency+ against the value of +priority+
    # attribute.
    #
    # @param [Money::Currency] other_currency The currency to compare to.
    #
    # @return [-1,0,1] -1 if less than, 0 is equal to, 1 if greater than
    #
    # @example
    #   c1 = Money::Currency.new(:usd)
    #   c2 = Money::Currency.new(:jpy)
    #   c1 <=> c2 #=> 1
    #   c2 <=> c1 #=> -1
    #   c1 <=> c1 #=> 0
    def <=>(other_currency)
      self.priority <=> other_currency.priority
    end

    # Compares +self+ with +other_currency+ and returns +true+ if the are the
    # same or if their +key+ attributes match.
    #
    # @param [Money::Currency] other_currency The currency to compare to.
    #
    # @return [Boolean]
    #
    # @example
    #   c1 = Money::Currency.new(:usd)
    #   c2 = Money::Currency.new(:jpy)
    #   c1 == c1 #=> true
    #   c1 == c2 #=> false
    def ==(other_currency)
      self.equal?(other_currency) ||
      self.key == other_currency.key
    end

    # Compares +self+ with +other_currency+ and returns +true+ if the are the
    # same or if their +key+ attributes match.
    #
    # @param [Money::Currency] other_currency The currency to compare to.
    #
    # @return [Boolean]
    #
    # @example
    #   c1 = Money::Currency.new(:usd)
    #   c2 = Money::Currency.new(:jpy)
    #   c1.eql? c1 #=> true
    #   c1.eql? c2 #=> false
    def eql?(other_currency)
      self == other_currency
    end

    # Returns a Fixnum hash value based on the +key+ attribute in order to use
    # functions like & (intersection), group_by, etc.
    #
    # @return [Fixnum]
    #
    # @example
    #   Money::Currency.new(:usd).hash #=> 428936
    def hash
      key.hash
    end

    # Returns a human readable representation.
    #
    # @return [String]
    #
    # @example
    #   Money::Currency.new(:usd) #=> #<Currency id: 146, key: usd ...>
    def inspect
      "#<#{self.class.name} id: #{id}, key: #{key}, priority: #{priority}, symbol_first: #{symbol_first}, thousands_separator: #{thousands_separator}, html_entity: #{html_entity}, decimal_mark: #{decimal_mark}, name: #{name}, symbol: #{symbol}, subunit_to_unit: #{subunit_to_unit}, iso_code: #{iso_code}, iso_numeric: #{iso_numeric}, subunit: #{subunit}>"
    end

    # Returns a string representation corresponding to the upcase +key+
    # attribute.
    #
    # -–
    # DEV: key.to_s.upcase corresponds to iso_code but don't use ISO_CODE for consistency.
    #
    # @return [String]
    #
    # @example
    #   Money::Currency.new(:usd).to_s #=> "USD"
    #   Money::Currency.new(:eur).to_s #=> "EUR"
    def to_s
      key.to_s.upcase
    end

    # Conversation to +self+.
    #
    # @return [self]
    def to_currency
      self
    end


    # Returns currency symbol or iso code for currencies with no symbol.
    #
    # @return [String]
    def code
      symbol || iso_code
    end

    def symbol_first?
      !!@symbol_first
    end

    # Cache decimal places for subunit_to_unit values.  Common ones pre-cached.
    def self.decimal_places_cache
      @decimal_places_cache ||= {
        1 => 0,
        10 => 1,
        100 => 2,
        1000 => 3
      }
    end

    # The number of decimal places needed.
    #
    # @return [Integer]
    def decimal_places
      cache = self.class.decimal_places_cache
      places  = cache[subunit_to_unit]
      unless places
        places = calculate_decimal_places(subunit_to_unit)
        cache[subunit_to_unit] = places
      end
      places
    end

    # If we need to figure out how many decimal places we need we
    # use repeated integer division.
    def calculate_decimal_places(num)
      return 0 if num == 1
      i = 1
      while num >= 10
        num /= 10
        i += 1 if num >= 10
      end
      i
    end
    private :calculate_decimal_places

  end
end
